### PR TITLE
[Qwen3-TTS] Serve Qwen3-TTS on Apple Silicon (MLX + Torch MPS)

### PR DESCRIPTION
**Pull request by AkazaAkane**
Runs Qwen3-TTS on Apple Silicon through two deliberately separate paths:

- **MLX**: the optimized native Apple backend, ported from [mlx-audio](https://github.com/Blaizzy/mlx-audio) without taking it as a dependency (`mlx-audio` requires `transformers>=5.14.0`; we pin `5.12.1`).
- **Torch MPS**: a conservative B=1 compatibility/reference path using the canonical upstream PyTorch talker and speech tokenizer.

Draft: **not ready to merge.** Opening it for review of the approach, and it may still want splitting. The commits are reviewable in order.

## Architecture direction

MLX is an optional optimized backend. PyTorch remains the canonical model semantics and owns device selection:

```text
Qwen3-TTS scheduler
├── torch
│   ├── cuda  SGLang-optimized serving
│   ├── mps   upstream Torch talker, eager/SDPA, B=1
│   └── cpu
└── mlx       native optimized Apple backend
```

This follows the Qwen3-ASR precedent: MPS is a Torch device, not a third model backend, and it bypasses the CUDA-specific execution core instead of accumulating MPS exceptions in it. The shared scheduler, request construction, prompt state, and result adapters remain unchanged.

The Torch MPS path replaces the SGLang CUDA-oriented generation modules with upstream `Qwen3TTSTalkerForConditionalGeneration`. Its `forward()` remains responsible for `DynamicCache`, feedback state, generation step, and groups 1..15 through the canonical code predictor. Only B=1 semantic sampling is implemented locally. This preserves a canonical Torch model suitable for future `torch.export` / `coreai-torch` work.

## Torch MPS qualification boundary

The initial path is intentionally narrow:

- one running/queued request;
- eager execution with `torch_native` attention and PyTorch sampling;
- no CUDA graph, compile, radix cache, overlap schedule, or chunked prefill;
- final/non-streaming official Torch speech-tokenizer decode;
- streaming requests are rejected explicitly.

CUDA and MLX behavior are unchanged.

## Commits

1. `[MLX]` Dispatch the MLX worker runner by architecture (registry; ASR behaviour unchanged)
2. `[Qwen3-TTS]` Talker + code predictor
3. `[Qwen3-TTS]` Speech tokenizer, speaker encoder, DSP
4. `[Qwen3-TTS]` Native prompt construction, CLI, playground
5. `[Qwen3-TTS]` Scheduler + vocoder-stage wiring
6. `[Qwen3-TTS]` Separate model backend selection from Torch device selection
7. `[Qwen3-TTS]` Add canonical Torch MPS serving path

MLX activation is `SGLANG_USE_MLX=1` plus importable MLX. Torch MPS is selected through the normal Torch backend with an MPS device.

## Two upstream bugs found

**The code2wav decoder is sliding-window attention, window 72.** The official config exposes `layer_types` as a computed property, so it never appears in `config.json`, and mlx-audio stores the window without ever masking with it. Invisible in short chunks; diverges once a pass exceeds 72 frames, which is the voice-cloning path where reference and generated codes decode in one pass. Measured divergence at 96 frames is **0.59** on a waveform in [-1, 1]. Honouring the window also makes chunked decoding exact at left context >= 71 rather than approximate.

**Codec tables have a load-order trap.** mlx-audio recomputes them in an overridden `update()` that `load_weights` never calls, so its loader patches around it. Folding `embedding_sum / cluster_usage` at sanitize removes the trap.

## Verified

- **MLX talker bit-identical to mlx-audio** on `Qwen3-TTS-12Hz-0.6B-CustomVoice`: all 402 tensors load strict; prefill hidden/logits, the 16-group frame, summed codec embedding, and post-feedback step all have `max|diff| == 0.0` in bf16 at B=1 and B=2.
- **Encoder matches the official Torch `MimiModel`** on 100% of codes at 2s/8s/24s.
- Speaker encoder is bit-identical; the mel filterbank is closer to librosa than mlx-audio's.
- Streamed MLX vocoder deltas concatenate to the one-shot decode; two interleaved streams each match being decoded alone.
- CustomVoice prompts generate speech codes rather than immediate EOS; preset voices diverge; an explicit language adds exactly one control position.
- MLX voice cloning runs end to end and writes a WAV via `examples/qwen3_tts_mlx_clone.py` and the Gradio playground.
- **Canonical Torch talker smoke** on the real 0.6B CustomVoice checkpoint: all 402 `talker.*` tensors strict-load into 905,788,672 parameters; no meta tensors remain; a two-token prefill followed by one decode produces finite `(1, 2, 3072)` logits and a complete `(1, 16)` codec frame.
- Current Qwen3-TTS unit suite: **381 passed, 5 skipped**. Focused Torch-MPS/backend checks: **12 passed**. Pre-commit passes on all files.

## Not verified

- **No real Apple Silicon / MPS run was possible on this Linux host.** The MPS runtime profile, upstream state machine, strict real-weight load, sampling contract, and final vocoder adapter are covered, but Metal operator support and end-to-end audio quality still require Mac qualification.
- **No end-to-end `sgl-omni serve` run, on any host.** Every stage is unit-tested and model components are checked against real weights, but the assembled pipeline is not.
- **No Apple Silicon timings.** MLX model checks ran on x86-64 CPU MLX (`mlx[cpu]`), which validates correctness but not Metal performance.
- `_rope_safe` guards a Metal `mx.fast.rope` bug that does not reproduce on CPU, so that path remains untested here.
- Still absent: Torch MPS streaming vocoder, Torch MPS B>1, Core AI export qualification, continuous MLX batching beyond SGLang's batched decode, incremental MLX streaming output, and an HTTP-backed playground.

Generated with AI coding assistance.

### CODE REVIEW

**Comment by AkazaAkane** [comment]
Added the canonical Torch MPS compatibility path in c7ea60f, following the Qwen3-ASR split: MPS stays under the Torch backend and bypasses the CUDA-oriented runner. The initial scope is B=1, eager/SDPA, upstream talker + DynamicCache/code predictor, and final-only official Torch vocoder decode. The full Qwen3-TTS unit suite passes (381 passed, 5 skipped), and the real 0.6B CustomVoice talker strict-load/CPU forward smoke produces a complete 16-group frame with no remaining meta tensors.

**Comment by Sunny-Island** [comment]
Hi @AkazaAkane, I have an M4 Mac mini with 16 GB unified memory and would like to help with Apple Silicon validation for this PR.

The PR description says real Apple Silicon/MPS and end-to-end serving have not been tested, while #1967 now marks Metal and serving E2E validation as done. Could you clarify the current status and which commit and launch configuration I should test?

If additional 16 GB Mac coverage would be useful, I can start with the 0.6B CustomVoice model, checking non-streaming HTTP audio generation on MLX and Torch/MPS with one request at a time. I can share setup steps, logs, and memory measurements across repeated requests, and help investigate any reproducible issues.

**Comment by AkazaAkane** [comment]
> Hi @AkazaAkane, I have an M4 Mac mini with 16 GB unified memory and would like to help with Apple Silicon validation for this PR.
> 
> The PR description says real Apple Silicon/MPS and end-to-end serving have not been tested, while #1967 now marks Metal and serving E2E validation as done. Could you clarify the current status and which commit and launch configuration I should test?
> 
> If additional 16 GB Mac coverage would be useful, I can start with the 0.6B CustomVoice model, checking non-streaming HTTP audio generation on MLX and Torch/MPS with one request at a time. I can share setup steps, logs, and memory measurements across repeated requests, and help investigate any reproducible issues.

Thanks for pointing it out. I tried Metal and serving E2E on my mac m5 24gb and it works. It would be helpful if you wanna help verifying it.
